service/interpreter/workflowImpl.go

```diff
@@ -763,28 +764,35 @@ func invokeStateExecute(
 		// this is not a problem because the signalWithStart will be very fast and highly available
 		unifiedClient := env.GetUnifiedClient()
 
-		var workflowId string
+		signalWithStartOn := env.GetSharedConfig().Api.WaitForStateCompletionMigration.SignalWithStartOn
 
-		if state.WaitForKey != nil {
-			workflowId = service.IwfSystemConstPrefix + executionContext.WorkflowId + "_" + state.StateId + "_" + *state.WaitForKey
-		} else {
-			workflowId = service.IwfSystemConstPrefix + executionContext.WorkflowId + "_" + *executionContext.StateExecutionId
+		// signalWithStart with legacy workflowId (containing parent workflowId)
+		if provider.GetBackendType() == service.BackendTypeCadence ||
+			(provider.GetBackendType() == service.BackendTypeTemporal && (signalWithStartOn == "old" || signalWithStartOn == "both")) {
+			workflowId := utils.GetWorkflowIdForWaitForStateExecution(provider.GetWorkflowInfo(ctx).FirstRunID, *executionContext.StateExecutionId, *state.WaitForKey, state.StateId)
+
+			err = signalWithStart(unifiedClient, workflowId)
+			if err != nil && !unifiedClient.IsWorkflowAlreadyStartedError(err) {
+				// WorkflowAlreadyStartedError is returned when the started workflow is closed and the signal is not sent
+				// panic will let the workflow task will retry until the signal is sent
+				panic(fmt.Errorf("failed to signal on completion %w", err))
+			}
 		}
 
-		err := unifiedClient.SignalWithStartWaitForStateCompletionWorkflow(
-			context.Background(),
-			uclient.StartWorkflowOptions{
-				ID:                       workflowId,
-				TaskQueue:                env.GetTaskQueue(),
-				WorkflowExecutionTimeout: 60 * time.Second, // timeout doesn't matter here as it will complete immediate with the signal
-			},
-			iwfidl.StateCompletionOutput{})
-		if err != nil && !unifiedClient.IsWorkflowAlreadyStartedError(err) {
-			// WorkflowAlreadyStartedError is returned when the started workflow is closed and the signal is not sent
-			// panic will let the workflow task will retry until the signal is sent
-			panic(fmt.Errorf("failed to signal on completion %w", err))
+		// signalWithStart with new workflowId (containing firstRunId)
+		if provider.GetBackendType() == service.BackendTypeTemporal && (signalWithStartOn == "both" || signalWithStartOn == "new") {
+			workflowId := utils.GetWorkflowIdForWaitForStateExecution(provider.GetWorkflowInfo(ctx).FirstRunID, *executionContext.StateExecutionId, *state.WaitForKey, state.StateId)
+
+			// Start WaitForStateCompletionWorkflow with a new name to ensure smooth transition
+			err = signalWithStart(unifiedClient, workflowId)
+			if err != nil && !unifiedClient.IsWorkflowAlreadyStartedError(err) {
+				// WorkflowAlreadyStartedError is returned when the started workflow is closed and the signal is not sent
+				// panic will let the workflow task will retry until the signal is sent
+				panic(fmt.Errorf("failed to signal on completion %w", err))
+			}
 		}
 	}
```

Comment:
Refactored this from two nested if statements into two separate ifs with more complicated expressions